types/src/blob/commitment.rs

```diff
@@ -0,0 +1,215 @@
+use std::io::Cursor;
+use std::num::NonZeroU64;
+
+use tendermint::{crypto, merkle};
+use tendermint_proto::v0_34::types::Blob as RawBlob;
+
+use crate::consts::appconsts;
+use crate::nmt::Namespace;
+use crate::{Error, Result, Share};
+
+/// create_commitment generates the share commitment for a given blob.
+/// See [Message layout rationale] and [Non-interactive default rules].
+///
+/// [Message layout rationale]: https://github.com/celestiaorg/celestia-specs/blob/e59efd63a2165866584833e91e1cb8a6ed8c8203/src/rationale/message_block_layout.md?plain=1#L12
+/// [Non-interactive default rules]: https://github.com/celestiaorg/celestia-specs/blob/e59efd63a2165866584833e91e1cb8a6ed8c8203/src/rationale/message_block_layout.md?plain=1#L36
+pub fn create_commitment(blob: &RawBlob) -> Result<Vec<u8>> {
+    let namespace = Namespace::new(blob.namespace_version as u8, &blob.namespace_id)?;
+    // TODO: we could probably lower amount of allocations there if
+    // Shares were just a wrapper over a raw data
+    let shares = split_blob_to_shares(blob)?;
+
+    // the commitment is the root of a merkle mountain range with max tree size
+    // determined by the number of roots required to create a share commitment
+    // over that blob. The size of the tree is only increased if the number of
+    // subtree roots surpasses a constant threshold.
+    let subtree_width = subtree_width(shares.len() as u64, appconsts::SUBTREE_ROOT_THRESHOLD);
+    let tree_sizes = merkle_mountain_range_sizes(shares.len() as u64, subtree_width);
+
+    let mut shares = shares;
+    let mut leaf_sets: Vec<Vec<Share>> = Vec::new();
+
+    for size in tree_sizes {
+        let left_shares = shares.split_off(size as usize);
+        leaf_sets.push(shares);
+        shares = left_shares;
+    }
+
+    // create the commitments by pushing each leaf set onto an nmt
+    let mut subtree_roots: Vec<Vec<u8>> = Vec::new();
+    for leaf_set in leaf_sets {
+        // create the nmt
+        let mut tree = crate::nmt::Nmt::new();
+        for leaf_share in leaf_set {
+            tree.push_leaf(
+                &leaf_share.to_vec(),
```

Comment:
Nevermind, i figured it out.

Reply:
> gives back a `Vec<&[u8]>`

it would have to be `Vec<Vec<u8>>` as shares aren't just 'a windows' over data

We surely can avoid share construction at this point as that's the only place where it's used, but I'm rather thinking if share shouldn't be implemented like in Go, so that it wraps raw `Vec<u8>` and provides some API around it reading it on the fly. Then it wouldn't need to allocate/free on `new` and `to_vec` (`into_inner`). + we could have `AsRef<u8>`